addons/website/static/src/js/editor/snippets.options.js

```diff
@@ -1027,6 +1030,61 @@ options.registry.anchor = options.Class.extend({
     },
 });
 
+options.registry.CookiesBar = options.registry.SnippetPopup.extend({
+    xmlDependencies: [
+        '/website/static/src/xml/website.cookies_bar.xml',
+    ],
+
+    //--------------------------------------------------------------------------
+    // Options
+    //--------------------------------------------------------------------------
+
+    /**
+     * Change the cookies bar layout.
+     *
+     * @see this.selectClass for parameters
+     */
+    selectLayout: function (previewMode, widgetValue, params) {
+        let websiteId;
+        this.trigger_up('context_get', {
+            callback: function (ctx) {
+                websiteId = ctx['website_id'];
+            },
+        });
+
+        const template = $(qweb.render(`website.cookies_bar.${widgetValue}`, {
+            websiteId: websiteId,
+        }));
+
+        const $content = this.$target.find('.s_popup_content');
+        const selectorToKeep = [
+            '.cookies_bar_text_button',
+            '.cookies_bar_text_policy',
+            '.cookies_bar_text_title',
+            '.cookies_bar_text_primary',
+            '.cookies_bar_text_secondary',
+        ];
+
+        if (this.text === undefined) {
+            this.text = [];
+        }
+
+        for (const selector of selectorToKeep) {
+            const $el = $content.find(selector);
+            if ($el.length) {
+                // save value before change, eg 'title' is not inside 'discrete' template
+                // but we want to preserve it in case of select another layout later
+                this.text[selector] = _.map($el, (el) => {
+                    return el.innerHTML;
+                });
+            }
+            template.find(selector).html(this.text[selector]);
+        }
```

Comment:
As @odony suggested indeed: instead of `innerHTML` -> `html`, use `childNodes` -> `appendChild` (wrong reason but worth mentioning: it should be 100x faster since using html requires parsing and recreation of DOM)

```javascript
this.allChildNodes = _.map($el, el => el.childNodes);
...
$template.find(selector).append(this.allChildNodes[selector]);
```

But in fact I am not sure it will be working since your use of `small` and `p` and ... etc. For example I wonder that if the user hits enter, it will create a new `p` which is probably not autorized in a `small` element and so might be extracted by smart browsers like Firefox ;) 

I wonder if each portion that you wanna keep can not be gather in a unique DIV then the code here would just be about moving that whole `div`? Instead of
```xml
<t t-name="default_stuff">Some text</t>
<small class="your_unique_selector">
    <t t-call="default_stuff"/>
</small>
```
Doing:
```xml
<t t-name="default_stuff"><div class="your_unique_selector">Some text</div></t>
<div class="small">
    <t t-call="default_stuff"/>
</div>
```
That way the unique selector is part of the t-call template and is not duplicated. It comes without any styling and the caller is responsible for the styling with a wrapping element. Then on layout change, it will just about moving the old div at the location of the new one. Not possible ?

Reply:
> But in fact I am not sure it will be working since your use of small and p and ... etc. For example I wonder that if the user hits enter, it will create a new p which is probably not autorized in a small element and so might be extracted by smart browsers like Firefox ;)

That's not how the editor handles its basic elements (the ones in the topbar, `p`, `pre`, `small`, `h1` etc..). It does not add a `p` on enter but a new tag. So an enter inside a `small` will create another `small` to go next line.
Note that Shift Enter will simply enter `<br/>`.

Everything is worked perfectly fine on both Chrome and Firefox, not sure what bug are we referring to.
Did a quick video of the behavior, [see here](https://drive.google.com/file/d/12YFFYLtogbHyM_hJG4hm2rYbMM2GaAUR/view).

Am I missing the point or is there any valid reason to change current implementation about moving texts?

Reply:
I am not trying to make you change code for no reason @rdeodoo, I like the feature. I am just trying to make this a possible merge for today :wink: 

- Hit enter four times in discrete cookie bar, the `<p/>` is duplicated four times with "primary" class and nothing visually happens, go to classic, go back to discrete, not the same as before because the `<p/>` are converted to `<br/>` somehow.
- Hit enter in popup mode, you have two `<p/>` with "secondary" class. Go back to classic mode, you have one `<p/>` with one `<br/>` inside... and going back to popup mode will not solve the problem, so the semantic construction of your layout was broken by just trying another option.
- Like the guessed example I gave, add a table in your element, it pops out the element since `<p/>`/`<span/>` cannot contain a table and is lost beyond repair on layout switching. Same for using a list which seems nice to have in a cookie bar policy.
- ... etc

All of this cover the same point I tried to explain above: layout switching should just be about **one container node** (_to take into account everything that might have been done inside_) to **move** (_to not reconstruct DOM on layout change which may lead to undesired behavior like JS binding losses (and is also slower)_).

Bonus: Also I don't think using the `small` tag make sense here, it should be using the `small` class since you just wanna say it is smaller, not give any semantic meaning to the text if I am not mistaken? Unfortunately, the editor is not smart enough to remove a small class if the user desires to do it but it will be the case for 14.0 and at worst the user can use the font-size stuff.

Also use `o_` as classes prefix please :wink:

Reply:
> Also use o_ as classes prefix please

:+1: 

 > Hit enter four times in discrete cookie bar, the `<p/>` is duplicated four times with "primary" class and nothing visually happens

This has nothing to do with my code (at least directly, indirectly I could've find another way to position stuff).
This is a limitation of css/editor. Normal enter is not working with `d-flex` and  `align-items-center`. It creates a new element (another `p`) but it is displayed next to the previous, so you visually don't see anything.
That something I wanted to discuss, see [here](https://github.com/odoo/odoo/compare/b3c95fa5c4ae470b9db080a5e325275b3cdcb581..3ced211957b0e5db35fd1edb97430cee1d7f7e0c#diff-24b0433302053618c5b742f3eded1d67L1063), but removed it from the PR at the end..

> Hit enter in popup mode, you have two `<p/>` with "secondary" class. Go back to classic mode, you have one `<p/>` with one `<br/>` inside... and going back to popup mode will not solve the problem, so the semantic construction of your layout was broken by just trying another option.
Like the guessed example I gave, add a table in your element, it pops out the element since `<p/>`/`<span/>` cannot contain a table and is lost beyond repair on layout switching. Same for using a list which seems nice to have in a cookie bar policy.

Ok, I got your point, everything of that I am aware of, purposely how I expected it to work.
Maybe bad design but that's what I chose to implement.
The goal of this was to preserve basic layouting as you can see [here](https://drive.google.com/file/d/12YFFYLtogbHyM_hJG4hm2rYbMM2GaAUR/view).

If we want to keep the whole stuff that was designed, I can do it. I think it will be overkill but whatever. I want discrete mode to remain discrete, just keep texts, not the whole layout that you constructed afterward in popup mode. Otherwise, when switching from the popup layout, that you fully customized, to discrete mode, it won't have much sense.
But I guess, if it bugged you, that 'losing' what you constructed when changing layout is not UX friendly.

Will try with your approach to keep everything.

Reply:
> I want discrete mode to remain discrete

Is not this the point of primary/secondary blocks ? You just keep primary if I remember well ? If not then maybe for discrete you can simply lose all text... but clearly not in preview mode :/ Just hovering an option should not make you lose everything.

>  I think it will be overkill but whatever.

The point is that doing it as I suggested results in **less and safer** code... and might prevent buggy behaviors (which are because of the editor, not your code... but your code runs inside the editor so we have to deal with it)... how is that overkill ?

Reply:
> Is not this the point of primary/secondary blocks ? You just keep primary if I remember well ?

Not really, the goal is the have a nice sentence self sufficient in discrete mode that you can customize.
In other mode, that sentence is kept and used, but also another whole paragraph is added to have more content with more explanation on the cookies bar. When you change the primary sentence, the change is reflected in every mode.

> how is that overkill ?

-> _I want discrete mode to remain discrete, just keep texts, not the whole layout that you constructed afterward in popup mode. Otherwise, when switching from the popup layout, that you fully customized, to discrete mode, it won't have much sense._

Currently adapting the code anyway so this one can move forwards